### PR TITLE
Print pod logs for failed scenarios

### PR DESCRIPTION
**Pull request by Sgitario**
Changes:
- Allow to configure the command to specify the output (whether a file or the console)
- Update OpenShiftTestExtension to run a list of actions when a scenario fails
- Add two actions to be run after a scenario failure: CopyLogsOnOpenShiftFailureActionImpl and PrintStatusOnOpenShiftFailureActionImpl.

### CODE REVIEW

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
+import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.io.OutputStream;
 import java.util.Arrays;
 import java.util.List;
-
-import static org.fusesource.jansi.Ansi.ansi;
+import java.util.function.BiConsumer;
```
Comment (Sgitario) [inline common/src/main/java/io/quarkus/ts/openshift/common/Command.java:14]: I used the same formatter than Quarkus ... should I revert this and use another? Don't want to create unnecessary changes everywhere.

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java [hunk 2]
```diff
@@ -22,35 +28,67 @@ public Command(List<String> command) {
...
-                .redirectErrorStream(true)
-                .command(command)
-                .directory(new File(".").getAbsoluteFile())
-                .start();
+            int result = process.waitFor();
+            if (result != 0) {
+                throw new RuntimeException(description + " failed (executed " + command + ", return code " + result + ")");
+            }
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
```
Comment (Ladicek) [inline common/src/main/java/io/quarkus/ts/openshift/common/Command.java:66]: Any particular reason to wrap exceptions like this?  Whomever calls `runAndWait` should be prepared to deal with interruption.

File: common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java [hunk 3]
```diff
@@ -450,6 +455,15 @@ public void handleTestExecutionException(ExtensionContext context, Throwable thr
         throw throwable;
     }
 
+    private void runOnFailureAction(ExtensionContext context, OnOpenShiftFailureAction action) {
+        try {
+            postProcessTestInstance(action, context);
```
Comment (Ladicek) [inline common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java:460]: This looks weird on the first sight, we should extract the injection logic into a separate, properly named method, so that it's clear what we do here.

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 1 above)
Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:14]: There's an import sort Maven plugin configured in this project, `mvn compile` should apply I think?

File: common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java (hunk 3 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java:460]: I agree with you, but we cannot change the postProcessTestInstance name... I can create another proper named method and make the postProcessTestInstance and runOnFailureAction to call to the new one if you prefer so.

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 2 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:66]: I will undo this and add the try/catch in the new callers.

File: common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java (hunk 3 above)
Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java:460]: That's what I meant :-)

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 2 above)
Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:66]: I think the callers can just propagate the exception, too. The generic "on failure action" interface should `throws Exception`, because it can execute arbitrary code.

That way, when error happens, you don't have a terribly long stack trace to go through.

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 1 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:14]: I don't mean only about sort import, but general formatter rules. This is the formatter I think we should follow: https://github.com/quarkusio/quarkus/tree/master/independent-projects/ide-config/src/main/resources

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 2 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:66]: Not a big fan of having an interface with "throws Exception", I prefer specific exceptions instead. 
On the other hand, these interfaces are to be called when the test scenario fails. So, I guess it's fine to state that the method might throw any exception and this exception will not hide the original root cause of the test failure (it will only print a log message about the exception). 
So, I will change the interface definition and document it properly.

Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java]: Changed.

File: common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java (hunk 3 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java]: Changed.

File: common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java [hunk 4]
```diff
@@ -38,16 +39,20 @@
...
 import java.util.Optional;
+import java.util.ServiceLoader;
 import java.util.stream.Stream;
 
 import static org.fusesource.jansi.Ansi.ansi;
 import static org.junit.platform.commons.util.AnnotationUtils.findAnnotatedFields;
 
-// TODO at this point, this class is close to becoming unreadable, and could use some refactoring
+// TODO at this point, this class is close to becoming unreadable, and could use some refactoring. 
+// Raised https://github.com/quarkus-qe/quarkus-openshift-test-suite/issues/108 for the refactoring.
```
Comment (Sgitario) [inline common/src/main/java/io/quarkus/ts/openshift/common/OpenShiftTestExtension.java:49]: As stated in the comment, I've raised the https://github.com/quarkus-qe/quarkus-openshift-test-suite/issues/108 issue to do this.

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 1 above)
Comment (rsvoboda) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:12]: yup, formater inclusion is on our todo, last time @Ladicek had some objections to Quarkus formatter, so we didn't push hard on it. But as more people will be looking into the project, it might be time to push it :)

Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:12]: The Eclipse formatter is crap and more often than not, I have to fight it to leave the code at least a little readable. I hate it with a passion, but feel free to use it on new code :-)

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 2 above)
Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java]: I personally think that if the "on failure" action should be able to "do anything", then it naturally needs to be able to "throw anything". See e.g. `java.util.concurrent.Callable` :-)

File: common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    public void execute() throws Exception {
+        String namespace = openShiftUtil.getNamespace();
+        for (Pod pod : openShiftUtil.getPods()) {
+            String podName = pod.getMetadata().getName();
+            new Command("oc", "logs", podName).outputToFile(getOutputFile(podName, namespace)).runAndWait();
+        }
+    }
+
+    static String getOutputFolder() {
+        return Config.get().getAsString(INSTANCES_LOGS_OUTPUT_DIRECTORY, DEFAULT_LOG_OUTPUT_DIRECTORY);
```
Comment (Ladicek) [inline common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java:30]: 1. You can also inject `Config` with `@TestResource`.
2. The config key must be `ts.*`. The config system will ignore keys that don't begin with `ts.`.

**Comment by Ladicek** [comment]
Just one comment, otherwise LGTM.

File: common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java (hunk 5 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java]: Changed

File: common/src/main/java/io/quarkus/ts/openshift/common/Command.java (hunk 1 above)
Comment (rsvoboda) [reply common/src/main/java/io/quarkus/ts/openshift/common/Command.java:12]: Whole codebase would get (re)formatted, I would go with https://github.com/quarkusio/quarkus/blob/master/build-parent/pom.xml#L459 way not to need to worry about everyone's IDE setup

File: common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import io.quarkus.ts.openshift.common.Command;
+import io.quarkus.ts.openshift.common.config.Config;
+import io.quarkus.ts.openshift.common.injection.TestResource;
+import io.quarkus.ts.openshift.common.util.OpenShiftUtil;
+
+import java.io.File;
+
+public class CopyLogsOnOpenShiftFailureActionImpl implements OnOpenShiftFailureAction {
+
+    private static final String INSTANCES_LOGS_OUTPUT_DIRECTORY = "instance-logs";
```
Comment (rsvoboda) [inline common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java:13]: @Ladicek wrote `The config key must be ts.*. The config system will ignore keys that don't begin with ts..`
You probably forgot about it.

README should be updated to include info about this property

**Review by rsvoboda — CHANGES_REQUESTED** [review]

File: common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java (hunk 6 above)
Comment (Ladicek) [reply common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java:13]: It's in the `Config` javadoc.

Comment (rsvoboda) [reply common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java:13]: exactly

Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/actions/CopyLogsOnOpenShiftFailureActionImpl.java]: Good point @rsvoboda . Added.

**Review by Ladicek — APPROVED** [review]

**Review by rsvoboda — APPROVED** [review]
